ministryofjustice/laa-apply-for-criminal-legal-aid
# CRIMAP-239 Preliminary approach to re-hydration

## Description of change
As part of the "returned application journey", on Apply we want to be able to select a returned application in the dashboard, and click a button "Update application" to edit it and re-submit.

The edition in reality is the creation of a brand new application, keeping all the entered information from the original (or parent) application, including same USN.

The provider is then presented with the `Check your answers` page for this new application, with everything pre-filled so they can change anything they are required to.

Then they can submit the application, which will create a new record in the datastore, with a new UUID, but same USN.

This is the first part of probably a few more PRs to finalise this re-hydration, taking a JSON payload and reconstructing a persisted (in Apply local DB) application in draft mode (`in_progress`), including relationships.

In this PR, a basic re-hydration of the top-level application attributes, and the `applicant` (including addresses if any) is included. Some other code tweaks to facilitate this, along with some renaming/refactoring.

## Link to relevant ticket
https://dsdmoj.atlassian.net/browse/CRIMAP-239

## Notes for reviewer

## Screenshots of changes (if applicable)
<img width="876" alt="Screenshot 2023-01-16 at 13 17 20" src="https://user-images.githubusercontent.com/687910/212687483-bc671c0f-8c6f-474a-bc79-39079fe29b7a.png">
<img width="716" alt="Screenshot 2023-01-16 at 13 17 38" src="https://user-images.githubusercontent.com/687910/212687494-e8a62b36-e90a-4c73-9e02-95f3490dd221.png">
<img width="667" alt="Screenshot 2023-01-16 at 13 17 50" src="https://user-images.githubusercontent.com/687910/212687510-c06efb7f-de47-48c9-9f8c-16292a4f9abf.png">


## How to manually test the feature
For the time being, until Review starts "returning" applications, this can be tested by manually updating the status of a submitted application in your local instance of the datastore. Set the status to `returned` and it should appear in the returned tab on Apply dashboard, and then you can re-hydrate it back to the Apply database.
